fms/utils/serialization.py

```diff
@@ -136,12 +168,353 @@ def load_state_dict(
+    rank: int
+        Rank of the current process
+    world_size: int
+        Total number of TP processes
+    prefix: str
+        Where to find the weight in the weights object
+    """
+    # Divide the weight matrix along the last dimension.
+    output_size_per_partition = module.weight.shape[1]
+    fms_weight_name = prefix + "weight"
+    if fms_weight_name in weight_name_map:
+        full_orig_name = weight_name_map[fms_weight_name]
+        tensor_slice = weights.get_slice(full_orig_name)
+        tensor = tensor_slice[
+            :,
+            (rank * output_size_per_partition) : (
+                (rank + 1) * output_size_per_partition
+            ),
+        ]
+        module.weight.copy_(tensor, non_blocking=True)
+    if module.bias is not None:
+        if rank == 0:
+            _copy_if_present_st(module.bias, weights, prefix + "bias", weight_name_map)
+        else:
+            module.bias.zero_()
+
+
+def _copy_embedding_st(
+    module: torch.nn.Module, weights, weight_name_map, rank, world_size, prefix
+):
+    """
+    This function copies the correct shard of the weights for a TP'd embedding module
+    according to the rank of the process and the world_size.
+
+    Args
+    ====
+    module: torch.nn.Module
+        Module that has had TP applied
+    weights:
+        safetensors weights object
+    rank: int
+        Rank of the current process
+    world_size: int
+        Total number of TP processes
+    prefix: str
+        Where to find the weight in the weights object
+    """
+    # Divide the weight matrix along the last dimension.
+    output_size_per_partition = module.weight.shape[1]
+    fms_weight_name = prefix + "weight"
+    if fms_weight_name in weight_name_map:
+        full_orig_name = weight_name_map[fms_weight_name]
+        tensor_slice = weights.get_slice(full_orig_name)
+        tensor = tensor_slice[
+            :,
+            (rank * output_size_per_partition) : (
+                (rank + 1) * output_size_per_partition
+            ),
+        ]
+        module.weight.copy_(tensor, non_blocking=True)
+
+
+def _copy_if_present_st(parameter, model_weights, fms_weight_name, weight_name_map):
+    if fms_weight_name in weight_name_map:
+        full_orig_name = weight_name_map[fms_weight_name]
+        parameter.copy_(model_weights.get_tensor(full_orig_name), non_blocking=True)
+
+
+def _load_safetensors_checkpoint_impl(
+    layer: torch.nn.Module,
+    model_weights,
+    weight_name_map,
+    rank=0,
+    world_size=1,
+    prefix="",
+):
+    # If we're on a leaf module and no TP has happened just copy the weights
+    if len(list(layer.children())) == 0:
+        for name, parameter in layer.named_parameters():
+            full_fms_name = prefix + name
+            _copy_if_present_st(
+                parameter, model_weights, full_fms_name, weight_name_map
+            )
+
+    for name, child in layer.named_children():
+        if isinstance(child, TPFeedForwardBlock):
```

Comment:
I think we should probably make a TPModule trait/class and an abstract function for this code. Each TP module can implement instead of having a list of isinstance checks. The `@abstract` should also help enforce that we don't forget to add it when adding any new TP modules.

Reply:
sure, I can do that as part of this PR

Reply:
I am very disappointed that traits are not supported as a thing in python... in any case, adding base abstract classes

Reply:
BTW, doesn't have to go in this PR, but at some point I think we'll also want to refactor this code to use the TPModule abstraction:
https://github.com/foundation-model-stack/foundation-model-stack/blob/main/fms/utils/tp_wrapping.py#L17